compiler/ast.nim

```diff
@@ -761,7 +761,62 @@ type
     itemId*: ItemId
   PIdObj* = ref TIdObj
 
-  PNode* = ref TNode
+  #----------------------------------------------------------------------------
+  # DOD AST Draft - Start
+  #----------------------------------------------------------------------------
+
+  # Actually implementing, would go something like:
+  # 1. repeat this pattern for P/TSym and P/TType
+  # 2. drop old P/TNode, P/TSym, and P/TType
+  # 3. implement inline accessors to replace fields
+  #    a. bash compiler until it compiles
+  # 5. discover zany uses of the P/TNode and friends...
+  #    a. bash compiler until it compiles
+  # 6. run all the tests
+  #    a. bash compiler until tests pass
+
+  PNode* = object
+    ## initial conversion point of AST to a data oriented design; rename to
+    ## `PNode` and swap out the old `PNode` for this
+    id: NodeId
+    state: State
+
+  NodeData = object
+    ## bare minimum data we need to know about every node
+    kind: TNodeKind ## presently the same as the nim node
+
+  AstTree = OrderedTableRef[NodeId, TNodeSeq]
+    ## store the tree structure for nodes, this is effectively `sons`
+    ## xxx: in here or separately we'll need to account for modules & fragments
+
+  State = ref object
+    nodeList: NodeList        ## each node, NodeId is their sequence index
+    nodeFlag: seq[TNodeFlags] ## flags for each node, rarely accessed bloat
+    nodeInf: seq[TLineInfo]   ## info on a per node basis
+
+    astData: AstTree          ## actual tree structure for the various AST
+
+    # sparse data, not all nodes have these
+    nodeSym: OrderedTableRef[NodeId, PSym]         ## symbols
```

Comment:
Why use table ref in addition to the state ref? We would be doing triple indirection in order to get the necessary data - (1) unref state, (2) unref table, (3) go to the table data store.

Reply:
Was bashing at something that wasn't working at the time, didn't revert.